jquery.overlay.js

```diff
@@ -211,23 +216,40 @@
           // Style attribute's string
           style = 'background-color:' + strategy.css['background-color'];
 
-          text.contents().each(function () {
-            var text, html, str, prevIndex;
-            if (this.nodeType != Node.TEXT_NODE) return;
-            text = this.textContent;
+          if (this.allowOverlapping) {
+            var textContent, prevIndex, str, html;
+            textContent = text.html()
             html = '';
             for (prevIndex = match.lastIndex = 0;; prevIndex = match.lastIndex) {
-              str = match.exec(text);
+              str = match.exec(textContent);
               if (!str) {
-                if (prevIndex) html += escape(text.substr(prevIndex));
+                if (prevIndex) html += textContent.substr(prevIndex);
                 break;
               }
               str = str[0];
-              html += escape(text.substr(prevIndex, match.lastIndex - prevIndex - str.length));
-              html += '<span style="' + style + '">' + escape(str) + '</span>';
+              html += textContent.substr(prevIndex, match.lastIndex - prevIndex - str.length);
+              html += '<span style="' + style + '">' + str + '</span>';
             };
-            if (prevIndex) $(this).replaceWith(html);
-          });
+            if (prevIndex) text.html(html);
+          } else {
+            text.contents().each(function () {
+              var text, html, str, prevIndex;
+              if (this.nodeType != Node.TEXT_NODE) return;
+              text = this.textContent;
+              html = '';
+              for (prevIndex = match.lastIndex = 0; ; prevIndex = match.lastIndex) {
+                str = match.exec(text);
+                if (!str) {
+                  if (prevIndex) html += escape(text.substr(prevIndex));
+                  break;
+                }
+                str = str[0];
+                html += escape(text.substr(prevIndex, match.lastIndex - prevIndex - str.length));
```

Comment:
Looks much more reasonable haha (note that you have duplicate semicolons in there somewhere too). I'd say just add a bit more documentation on what's going on -- this type of code really presupposes that a reader has a lot of context about the data structures / data being manipulated in their brain's RAM (so to speak), so a little commenting goes a long way. 

Just as a point of comparison, pieces of code like our ruby code for Filters that uses dependency injection really heavily is a bit better at self-documenting provided that method names are really clear. One tradeoff there is that it can be tougher to track specific behaviors down since it's more likely to live in multiple files.